src/main/java/rx/internal/producers/ProducerArbiter.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2015 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package rx.internal.producers;
+
+import rx.*;
+
+/**
+ * Producer that allows changing an underlying producer atomically and correctly resume with the accumulated
+ * requests.
+ */
+public final class ProducerArbiter implements Producer {
+    long requested;
+    Producer currentProducer;
+ 
+    boolean emitting;
```

Comment:
Yeah I noticed the naming pattern from your blogs and I think instead of a *queue drainer* and an *emitter* they are better described as both queue drainers except one is aysnc biased and the other is sync biased.